File: permuta_policial/lib/core/utils/salary_calculator.dart
// /lib/core/utils/salary_calculator.dart

import '../models/work_day.dart';
import '../models/salary.dart';

class SalaryCalculator {
  /// Calcula o número de etapas para um dia baseado nas horas trabalhadas
  /// [totalHours] - Total de horas do dia
  /// [etapaRule] - Regra de etapas ('per_6h', 'per_8h', 'fixed')
  /// [customValue] - Valor customizado (para regras fixed)
  /// Retorna o número de etapas (sempre arredondado para baixo)
  static int calcEtapasForDay(
    double totalHours, {
    String etapaRule = 'per_6h',
    double? customValue,
  }) {
    if (totalHours <= 0) return 0;

    switch (etapaRule) {
      case 'per_6h':
        return (totalHours / 6).floor();
      case 'per_8h':
        return (totalHours / 8).floor();
      case 'fixed':
        return customValue != null ? customValue.floor() : 0;
      default:
        // Default: 1 etapa a cada 6 horas
        return (totalHours / 6).floor();
    }
  }

  /// Calcula minutos de sobreposição entre dois intervalos
  static int computeOverlapMinutes(
    DateTime start1,
    DateTime end1,
    DateTime start2,
    DateTime end2,
  ) {
    final overlapStart = start1.isAfter(start2) ? start1 : start2;
    final overlapEnd = end1.isBefore(end2) ? end1 : end2;

    if (overlapStart.isAfter(overlapEnd) || overlapStart.isAtSameMomentAs(overlapEnd)) {
      return 0;
    }

    return overlapEnd.difference(overlapStart).inMinutes;
  }

  /// Tabela progressiva mensal tradicional (Lei 15.270/2025).
  static double _calcIRPFTradicional(double baseCalculo) {
    if (baseCalculo <= 2428.80) return 0;
    if (baseCalculo <= 2826.65) return baseCalculo * 0.075 - 182.16;
    if (baseCalculo <= 3751.05) return baseCalculo * 0.15 - 394.16;
    if (baseCalculo <= 4664.68) return baseCalculo * 0.225 - 675.49;
    return baseCalculo * 0.275 - 908.73;
  }

  /// Calcula IRPF mensal conforme regras de 2026 (Lei 15.270/2025).
  ///
  /// [baseCalculo] — (salário_base + horas_extras) − previdência
  static double calcIRPF(double baseCalculo) {
    if (baseCalculo <= 0) return 0;
    if (baseCalculo <= 5000) return 0;

    final irpfTradicional = _calcIRPFTradicional(baseCalculo).clamp(0.0, double.infinity);
    var irpf = irpfTradicional;

    if (baseCalculo <= 7350) {
      final redutor = 978.62 - (0.133145 * baseCalculo);
      final desconto = redutor.clamp(0.0, irpfTradicional);
      irpf = irpfTradicional - desconto;
    }

    if (baseCalculo > 50000) {
      final irpfmMinimo = baseCalculo * 0.10;
      irpf = irpf > irpfmMinimo ? irpf : irpfmMinimo;
    }

    return ((irpf * 100).round() / 100).clamp(0.0, double.infinity);
  }

  /// Calcula horas de um intervalo considerando timezone e turnos que cruzam meia-noite
  /// [startTime] - Início do intervalo
  /// [endTime] - Fim do intervalo
  /// Retorna horas (decimal)
  static double calcIntervalHours(DateTime startTime, DateTime endTime) {
    Duration diff = endTime.difference(startTime);
    
    if (diff.isNegative) {
      final nextDay = endTime.add(const Duration(days: 1));
      diff = nextDay.difference(startTime);
    }

    var hours = diff.inMinutes / 60.0;
    if (hours > 24) {
      hours = 24;
    }
    
    return ((hours * 100).round()) / 100;
  }

  /// Calcula o salário do mês completo
  /// [workDays] - Array de dias trabalhados
  /// [settings] - Configurações de salário do usuário
  /// [month] - Mês (1-12)
  /// [year] - Ano
  /// Retorna o resultado do cálculo
  static Map<String, dynamic> calcSalaryMonth(
    List<WorkDay> workDays,
    SalarySettings settings,
    int month,
    int year,
  ) {
    // Calcula dias do mês
    final daysInMonth = DateTime(year, month + 1, 0).day;
    var cargaHorariaMes = daysInMonth * (settings.cargaHorariaDia);

    // Totais do mês
    double totalHoras = 0;
    int totalEtapas = 0;
    int diasTrabalhados = 0;
    int diasFerias = 0;
    double horasAbatimento = 0;

    // Processa cada dia
    for (var day in workDays) {
      final dayType = day.tipo;

      if (dayType == 'ferias') {
        diasFerias++;
        continue; // Férias não contam horas/etapas
      }

      if (dayType == 'folga') {
        continue; // Folga não conta
      }

      final isAbatimento =
          day.flagAbatimento || dayType == 'abatimento' || dayType == 'atestado';

      if (isAbatimento) {
        final horasAbat = day.totalHours > 0 ? day.totalHours : settings.abatimentoHoras;
        horasAbatimento += horasAbat;
        continue;
      }

      // Calcula horas do dia
      double horasDia = 0;

      if (day.totalHours > 0) {
        horasDia = day.totalHours;
      }

      if (dayType == 'normal' && horasDia <= 0) {
        continue;
      }

      final etapaRule = day.etapaRuleOverride ?? settings.etapaRule;
      final etapasDia = calcEtapasForDay(horasDia, etapaRule: etapaRule);

      totalHoras += horasDia;
      totalEtapas += etapasDia;
      diasTrabalhados++;
    }

    cargaHorariaMes = (cargaHorariaMes - horasAbatimento).clamp(0.0, double.infinity);

    // Calcula horas extras
    final horasExtras = (totalHoras - cargaHorariaMes).clamp(0.0, double.infinity);

    // Valores monetários
    final valorEtapas = totalEtapas * settings.etapaValue;
    final valorHorasExtras = horasExtras * settings.valorHoraExtra;

    // VA: não creditar se houver dias de férias no mês
    final valeAlimentacao = diasFerias > 0 ? 0 : settings.valeAlimentacao;

    // Outras vantagens (ex: substituição)
    final outrasVantagens = settings.outrasVantagens;

    // Salário (sem VA) e total de recebimentos
    final salarioBrutoFolha = settings.salarioBase + valorHorasExtras + outrasVantagens;
    final salarioBruto = salarioBrutoFolha + valorEtapas;
    final totalRecebimentos = salarioBruto + valeAlimentacao;

    final basePrevidencia = settings.salarioBase + valorHorasExtras;
    final descontoPrevidencia = basePrevidencia * settings.previdenciaAliquota;

    final baseIRPF = (basePrevidencia - descontoPrevidencia).clamp(0.0, double.infinity);
    final descontoIRPF = calcIRPF(baseIRPF);

    final descontoConsignados = settings.descontoConsignados ?? 0;
    final outrosDescontos = settings.outrosDescontos;

    final salarioLiquidoFolha =
        salarioBrutoFolha - descontoPrevidencia - descontoIRPF - descontoConsignados - outrosDescontos;
    final vaMaisEtapas = valeAlimentacao + valorEtapas;
    final salarioLiquido = salarioLiquidoFolha + vaMaisEtapas;

    return {
      'total_horas': (totalHoras * 100).round() / 100,
      'horas_abatimento': (horasAbatimento * 100).round() / 100,
      'carga_horaria_mes': (cargaHorariaMes * 100).round() / 100,
      'horas_extras': (horasExtras * 100).round() / 100,
      'total_etapas': totalEtapas,
      'valor_etapas': (valorEtapas * 100).round() / 100,
      'vale_alimentacao': (valeAlimentacao * 100).round() / 100,
      'outras_vantagens': (outrasVantagens * 100).round() / 100,
      'valor_horas_extras': (valorHorasExtras * 100).round() / 100,
      'salario_base': (settings.salarioBase * 100).round() / 100,
      'salario_bruto': (salarioBruto * 100).round() / 100,
      'total_recebimentos': (totalRecebimentos * 100).round() / 100,
      'base_irpf': (baseIRPF * 100).round() / 100,
      'desconto_previdencia': (descontoPrevidencia * 100).round() / 100,
      'desconto_irpf': (descontoIRPF * 100).round() / 100,
      'desconto_consignados': (descontoConsignados * 100).round() / 100,
      'outros_descontos': (outrosDescontos * 100).round() / 100,
      'salario_liquido': (salarioLiquido * 100).round() / 100,
      'salario_liquido_folha': (salarioLiquidoFolha * 100).round() / 100,
      'va_mais_etapas': (vaMaisEtapas * 100).round() / 100,
      'dias_trabalhados': diasTrabalhados,
      'dias_ferias': diasFerias,
    };
  }
}

File: backend_js/src/modules/work/salary-calculator.js
  // /src/modules/work/salary-calculator.js

  const { DateTime } = require('luxon');

  /**
   * Calcula o número de etapas para um dia baseado nas horas trabalhadas
   * @param {number} totalHours - Total de horas do dia
   * @param {string} etapaRule - Regra de etapas ('per_6h', 'per_8h', 'fixed_X')
   * @param {number} customValue - Valor customizado (para regras fixed)
   * @returns {number} - Número de etapas (sempre arredondado para baixo)
   */
  function calcEtapasForDay(totalHours, etapaRule = 'per_6h', customValue = null) {
    if (!totalHours || totalHours <= 0) return 0;

    switch (etapaRule) {
      case 'per_6h':
        return Math.floor(totalHours / 6);
      case 'per_8h':
        return Math.floor(totalHours / 8);
      case 'fixed':
        return customValue ? Math.floor(customValue) : 0;
      default:
        // Default: 1 etapa a cada 6 horas
        return Math.floor(totalHours / 6);
    }
  }

  /**
   * Calcula minutos de sobreposição entre dois intervalos
   * @param {Date|string} start1 - Início do intervalo 1
   * @param {Date|string} end1 - Fim do intervalo 1
   * @param {Date|string} start2 - Início do intervalo 2
   * @param {Date|string} end2 - Fim do intervalo 2
   * @returns {number} - Minutos de sobreposição
   */
  function computeOverlapMinutes(start1, end1, start2, end2) {
    const s1 = DateTime.fromJSDate(new Date(start1));
    const e1 = DateTime.fromJSDate(new Date(end1));
    const s2 = DateTime.fromJSDate(new Date(start2));
    const e2 = DateTime.fromJSDate(new Date(end2));

    const overlapStart = s1 > s2 ? s1 : s2;
    const overlapEnd = e1 < e2 ? e1 : e2;

    if (overlapStart >= overlapEnd) {
      return 0;
    }

    return Math.round(overlapEnd.diff(overlapStart, 'minutes').minutes);
  }

  /**
   * Tabela progressiva mensal tradicional (Lei 15.270/2025 — vigência 2026).
   * @param {number} baseCalculo
   * @returns {number}
   */
  function calcIRPFTradicional(baseCalculo) {
    if (baseCalculo <= 2428.80) {
      return 0;
    }
    if (baseCalculo <= 2826.65) {
      return baseCalculo * 0.075 - 182.16;
    }
    if (baseCalculo <= 3751.05) {
      return baseCalculo * 0.15 - 394.16;
    }
    if (baseCalculo <= 4664.68) {
      return baseCalculo * 0.225 - 675.49;
    }
    return baseCalculo * 0.275 - 908.73;
  }

  /**
   * Calcula IRPF mensal conforme regras de 2026 (Lei 15.270/2025).
   *
   * Base de cálculo: (salário_base + horas_extras) − previdência — sem VA, etapas ou outras vantagens.
   *
   * Faixas:
   * - Até R$ 5.000: isento
   * - R$ 5.000,01 a R$ 7.350: tabela tradicional − redutor progressivo
   * - Acima de R$ 7.350: tabela tradicional integral
   * - Acima de R$ 50.000: trava IRPFM (alíquota mínima de 10% sobre a base)
   *
   * @param {number} baseCalculo
   * @returns {number}
   */
  function calcIRPF(baseCalculo) {
    if (baseCalculo <= 0) {
      return 0;
    }

    if (baseCalculo <= 5000) {
      return 0;
    }

    const irpfTradicional = Math.max(0, calcIRPFTradicional(baseCalculo));
    let irpf = irpfTradicional;

    if (baseCalculo <= 7350) {
      const redutor = 978.62 - (0.133145 * baseCalculo);
      const desconto = Math.max(0, Math.min(redutor, irpfTradicional));
      irpf = irpfTradicional - desconto;
    }

    if (baseCalculo > 50000) {
      const irpfmMinimo = baseCalculo * 0.10;
      irpf = Math.max(irpf, irpfmMinimo);
    }

    return Math.max(0, Math.round(irpf * 100) / 100);
  }

  /**
   * Calcula o salário do mês completo
   * @param {Object} params - Parâmetros de cálculo
   * @param {Array} workDays - Array de dias trabalhados
   * @param {Object} settings - Configurações de salário do usuário
   * @param {number} month - Mês (1-12)
   * @param {number} year - Ano
   * @returns {Object} - Resultado do cálculo
   */
  function calcSalaryMonth(workDays, settings, month, year) {
    const timezone = 'America/Sao_Paulo';
    
    // Calcula dias do mês
    const daysInMonth = DateTime.fromObject({ year, month, day: 1 }, { zone: timezone }).daysInMonth;
    let cargaHorariaMes = daysInMonth * (settings.carga_horaria_dia || 5.70);

    // Totais do mês
    let totalHoras = 0;
    let totalEtapas = 0;
    let diasTrabalhados = 0;
    let diasFerias = 0;
    let horasAbatimento = 0;

    // Processa cada dia
    workDays.forEach(day => {
      const dayType = day.tipo || 'normal';
      
      if (dayType === 'ferias') {
        diasFerias++;
        return; // Férias não contam horas/etapas
      }

      if (dayType === 'folga') {
        return; // Folga não conta
      }

      const isAbatimento =
        day.flag_abatimento || dayType === 'abatimento' || dayType === 'atestado';

      if (isAbatimento) {
        const horasAbat =
          day.total_hours && day.total_hours > 0
            ? parseFloat(day.total_hours)
            : settings.abatimento_horas || 5.70;
        horasAbatimento += horasAbat;
        return;
      }

      // Calcula horas do dia
      let horasDia = 0;
      
      if (day.total_hours && day.total_hours > 0) {
        horasDia = parseFloat(day.total_hours);
      }

      // Dia normal sem horas não entra no cálculo
      if (dayType === 'normal' && horasDia <= 0) {
        return;
      }

      // Calcula etapas do dia
      const etapaRule = day.etapa_rule_override || settings.etapa_rule || 'per_6h';
      const etapasDia = calcEtapasForDay(horasDia, etapaRule);

      totalHoras += horasDia;
      totalEtapas += etapasDia;
      diasTrabalhados++;
    });

    cargaHorariaMes = Math.max(0, cargaHorariaMes - horasAbatimento);

    // Calcula horas extras
    const horasExtras = Math.max(0, totalHoras - cargaHorariaMes);

    // Valores monetários
    const valorEtapas = totalEtapas * (settings.etapa_value || 11.00);
    const valorHorasExtras = horasExtras * (settings.valor_hora_extra || 44.00);
    
    // VA: não creditar se houver dias de férias no mês
    const valeAlimentacao = diasFerias > 0 ? 0 : (settings.vale_alimentacao || 426.00);

    // Outras vantagens (ex: substituição)
    const outrasVantagens = settings.outras_vantagens || 0;

    // Salário (sem VA) e total de recebimentos
    const salarioBrutoFolha =
      (settings.salario_base || 0) + valorHorasExtras + outrasVantagens;
    const salarioBruto = salarioBrutoFolha + valorEtapas;
    const totalRecebimentos = salarioBruto + valeAlimentacao;

    // Previdência: salário base + horas extras (VA, etapas e outras vantagens ficam de fora)
    const basePrevidencia = (settings.salario_base || 0) + valorHorasExtras;
    const descontoPrevidencia = basePrevidencia * (settings.previdencia_aliquota || 0.14);

    // IRPF: apenas (salário_base + horas_extras) − previdência
    const baseIRPF = Math.max(0, basePrevidencia - descontoPrevidencia);
    const descontoIRPF = calcIRPF(baseIRPF);

    const descontoConsignados = settings.desconto_consignados || 0;
    const outrosDescontos = settings.outros_descontos || 0;

    const salarioLiquidoFolha =
      salarioBrutoFolha - descontoPrevidencia - descontoIRPF - descontoConsignados - outrosDescontos;
    const vaMaisEtapas = valeAlimentacao + valorEtapas;
    const salarioLiquido = salarioLiquidoFolha + vaMaisEtapas;

    return {
      total_horas: Math.round(totalHoras * 100) / 100,
      horas_abatimento: Math.round(horasAbatimento * 100) / 100,
      carga_horaria_mes: Math.round(cargaHorariaMes * 100) / 100,
      horas_extras: Math.round(horasExtras * 100) / 100,
      total_etapas: totalEtapas,
      valor_etapas: Math.round(valorEtapas * 100) / 100,
      vale_alimentacao: Math.round(valeAlimentacao * 100) / 100,
      outras_vantagens: Math.round(outrasVantagens * 100) / 100,
      valor_horas_extras: Math.round(valorHorasExtras * 100) / 100,
      salario_base: Math.round((settings.salario_base || 0) * 100) / 100,
      salario_bruto: Math.round(salarioBruto * 100) / 100,
      total_recebimentos: Math.round(totalRecebimentos * 100) / 100,
      base_irpf: Math.round(baseIRPF * 100) / 100,
      desconto_previdencia: Math.round(descontoPrevidencia * 100) / 100,
      desconto_irpf: Math.round(descontoIRPF * 100) / 100,
      desconto_consignados: Math.round(descontoConsignados * 100) / 100,
      outros_descontos: Math.round(outrosDescontos * 100) / 100,
      salario_liquido: Math.round(salarioLiquido * 100) / 100,
      salario_liquido_folha: Math.round(salarioLiquidoFolha * 100) / 100,
      va_mais_etapas: Math.round(vaMaisEtapas * 100) / 100,
      dias_trabalhados: diasTrabalhados,
      dias_ferias: diasFerias
    };
  }

  /**
   * Calcula horas de um intervalo considerando timezone e turnos que cruzam meia-noite
   * @param {string|Date} startTime - Início do intervalo
   * @param {string|Date} endTime - Fim do intervalo
   * @param {string} timezone - Timezone (default: America/Sao_Paulo)
   * @returns {number} - Horas (decimal)
   */
  function calcIntervalHours(startTime, endTime, timezone = 'America/Sao_Paulo') {
    const start = DateTime.fromJSDate(new Date(startTime), { zone: timezone });
    let end = DateTime.fromJSDate(new Date(endTime), { zone: timezone });

    // Se end <= start, assume cruzamento de meia-noite
    if (end <= start) {
      end = end.plus({ days: 1 });
    }

    let hours = end.diff(start, 'hours').hours;
    // Limita a 24h por intervalo (proteção contra dados inválidos)
    if (hours > 24) {
      hours = 24;
    }

    return Math.round(hours * 100) / 100;
  }

  /**
   * Divide intervalos que cruzam meia-noite em múltiplos dias
   * @param {string|Date} startTime - Início do intervalo
   * @param {string|Date} endTime - Fim do intervalo
   * @param {string} timezone - Timezone
   * @returns {Array} - Array de intervalos por dia [{ date, start, end, hours }]
   */
  function splitIntervalByDay(startTime, endTime, timezone = 'America/Sao_Paulo') {
    const start = DateTime.fromJSDate(new Date(startTime), { zone: timezone });
    const end = DateTime.fromJSDate(new Date(endTime), { zone: timezone });

    const intervals = [];
    let currentStart = start;

    while (currentStart < end) {
      // Fim do dia atual (meia-noite)
      const dayEnd = currentStart.endOf('day');
      const intervalEnd = end < dayEnd ? end : dayEnd;

      const hours = calcIntervalHours(currentStart.toJSDate(), intervalEnd.toJSDate(), timezone);

      intervals.push({
        date: currentStart.toISODate(),
        start: currentStart.toJSDate(),
        end: intervalEnd.toJSDate(),
        hours: hours
      });

      // Próximo dia (meia-noite + 1 segundo)
      currentStart = intervalEnd.plus({ seconds: 1 });
    }

    return intervals;
  }

  module.exports = {
    calcEtapasForDay,
    calcIRPF,
    calcIRPFTradicional,
    calcSalaryMonth,
    computeOverlapMinutes,
    calcIntervalHours,
    splitIntervalByDay
  };


